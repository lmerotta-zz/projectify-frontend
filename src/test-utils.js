import React from "react";
import { render } from "@testing-library/react";
import { CacheProvider } from "@emotion/react";
import createCache from "@emotion/cache";

const cache = createCache({ key: "iodajdoiwanoawdni" });
cache.compat = true;

const AllTheProviders = ({ children }) => {
  return <CacheProvider value={cache}>{children}</CacheProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
