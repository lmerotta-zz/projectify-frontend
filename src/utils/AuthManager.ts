import { UserManager } from "oidc-client";

class AuthManager {
  private userManager: UserManager;
  constructor() {
    this.userManager = new UserManager({
      authority: process.env.REACT_APP_ACTIONS_URL,
      response_type: "code",
      client_id: process.env.REACT_APP_OAUTH_CLIENTID,
      redirect_uri: process.env.REACT_APP_OAUTH_REDIRECTURI,
      silent_redirect_uri: process.env.REACT_APP_OAUTH_SILENT_REDIRECTURI,
      scope: "email",
    });
  }

  async login(args: any) {
    await this.userManager.signinRedirect(args);
  }

  async isLoggedIn() {
    return await this.userManager.signinSilent({
      scope: "email",
    });
  }

  async getUser() {
    return await this.userManager.getUser();
  }

  async removeUser() {
    await this.userManager.removeUser();
  }

  async logout() {
    await this.userManager.signoutRedirect();
  }
}

export default new AuthManager();
