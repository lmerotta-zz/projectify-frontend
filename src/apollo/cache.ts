import { makeVar } from "@apollo/client";

type OAuthBag = {
  verifier: string;
  challenge: string;
  one_time_code?: string;
  access_token?: string;
  refresh_token?: string;
};

export const oauthBag = makeVar<OAuthBag>({
  verifier: "HjHMNAhDZZn9ln1wOHf6jD4oSGmw5OSilQV8n2jFwiA", // TODO: variabilize this !
  challenge: "UdFfYchv1a6tryk3-n_EsFm9CuBZqq5ICCv29SXTiiI", // TODO: variabilize this !
});
