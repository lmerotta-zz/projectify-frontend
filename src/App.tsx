import { captureException } from "@sentry/minimal";
import { isAuthenticated } from "apollo/local-state";
import { PrivateRoute } from "components";
import AppContainer from "modules/core/components/AppContainer";
import { SecurityPage } from "modules/security";
import { pluginStore } from "plugins";
import { useEffect } from "react";
import { PluginProvider } from "react-pluggable";
import { Route, Routes } from "react-router-dom";
import AuthManager from "utils/AuthManager";

const App = () => {
  useEffect(() => {
    setTimeout(async () => {
      try {
        await AuthManager.isLoggedIn();
        isAuthenticated(true);
      } catch (e) {
        await AuthManager.removeUser();
        captureException(e);
        isAuthenticated(false);
      }
    }, 0);
  }, []);

  return (
    <PluginProvider pluginStore={pluginStore}>
      <Routes>
        <Route
          path="/*"
          element={<PrivateRoute element={<AppContainer />} />}
        />
        <Route path="/security/*" element={<SecurityPage />} />
      </Routes>
    </PluginProvider>
  );
};

export default App;
