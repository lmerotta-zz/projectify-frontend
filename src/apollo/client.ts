import {
  ApolloClient,
  InMemoryCache,
  from,
  createHttpLink,
} from "@apollo/client";
import { RestLink } from "apollo-link-rest";
import { setContext } from "@apollo/client/link/context";
import i18next from "i18next";
import AuthManager from "utils/AuthManager";
import { isAuthenticated } from "./local-state";

const restLink = new RestLink({
  endpoints: {
    login: {
      uri: `${process.env.REACT_APP_ACTIONS_URL}/login`,
    },
  },
  credentials: "include",
});

const httpLink = createHttpLink({
  uri: process.env.REACT_APP_GRAPHQL_URL,
});

const authLink = setContext(async (_, { headers }) => {
  if (!isAuthenticated()) {
    return;
  }
  let user = await AuthManager.getUser();
  if (user === null || user.expired) {
    try {
      user = await AuthManager.isLoggedIn();
    } catch (e) {
      user = null;
    }
  }
  if (user !== null) {
    return {
      headers: {
        ...headers,
        Authorization: `Bearer ${user.id_token}`,
      },
    };
  }

  return {
    headers,
  };
});

let locale = "";

i18next.on("languageChanged", (lng) => {
  locale = lng;
});

const client = new ApolloClient({
  cache: new InMemoryCache(),
  headers: {
    "x-locale": locale,
  },
  link: from([restLink, authLink.concat(httpLink)]),
});

export default client;
