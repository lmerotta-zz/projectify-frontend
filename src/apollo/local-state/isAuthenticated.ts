import { makeVar } from "@apollo/client";

const isAuthenticated = makeVar<boolean | null>(null);

export default isAuthenticated;
