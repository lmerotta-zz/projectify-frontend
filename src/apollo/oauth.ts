import { makeVar } from "@apollo/client";

type OAuthBag = {
  verifier: string;
  challenge: string;
  one_time_code?: string;
  access_token?: string;
  refresh_token?: string;
};

export const oauthBag = makeVar<OAuthBag>({
  verifier: "HjHMNAhDZZn9ln1wOHf6jD4oSGmw5OSilQV8n2jFwiA", // used only for testing
  challenge: "UdFfYchv1a6tryk3-n_EsFm9CuBZqq5ICCv29SXTiiI", // used only for testing
});

/* istanbul ignore next */
const generateHashes = async () => {
  function base64URLEncode(str: string) {
    return btoa(str);
  }

  async function sha256(buffer: ArrayBuffer) {
    return await crypto.subtle.digest("SHA-256", buffer);
  }

  console.log(window);
  const verifierBuffer = window.crypto.getRandomValues(new Uint32Array(10));
  const verifier = base64URLEncode(verifierBuffer.join(""));

  const challengeBuffer = await sha256(verifierBuffer);
  const challenge = Array.prototype.map
    .call(new Uint8Array(challengeBuffer), (x) =>
      ("00" + x.toString(16)).slice(-2)
    )
    .join("");

  oauthBag({
    verifier,
    challenge,
  });
};

if (process.env.NODE_ENV !== "test") {
  generateHashes();
}
