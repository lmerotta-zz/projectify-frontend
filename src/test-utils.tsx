import { MockedProvider, MockedProviderProps } from "@apollo/client/testing";
import {
  CircularProgress,
  StyledEngineProvider,
  ThemeProvider,
} from "@mui/material";
import { render as realRender, RenderOptions } from "@testing-library/react";
import { ReactElement, ReactNode, Suspense } from "react";
import { MemoryRouter, MemoryRouterProps } from "react-router";
import { Route, Routes } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import theme from "theme";

function renderWithProviders(
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries"> & {
    routerProps?: MemoryRouterProps;
    graphqlProps?: MockedProviderProps;
  }
) {
  const Wrapper = ({ children }: { children?: ReactNode }) => {
    return (
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={theme}>
          <ToastContainer
            theme="colored"
            autoClose={3000}
            containerId="react-toastify"
          />
          <MockedProvider {...options?.graphqlProps} addTypename={false}>
            <MemoryRouter {...options?.routerProps}>
              <Suspense fallback={<CircularProgress />}>
                <Routes>
                  <Route
                    path="/*"
                    element={
                      <>
                        <h1>TESTSASATATAST</h1>
                        {children}
                      </>
                    }
                  />
                </Routes>
              </Suspense>
            </MemoryRouter>
          </MockedProvider>
        </ThemeProvider>
      </StyledEngineProvider>
    );
  };

  return realRender(ui, { wrapper: Wrapper, ...options });
}

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { renderWithProviders };
