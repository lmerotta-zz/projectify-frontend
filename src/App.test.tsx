import { isAuthenticated } from "apollo/local-state";
import App from "App";
import { renderWithProviders, waitFor } from "test-utils";
import AuthManager from "utils/AuthManager";

jest.mock("utils/AuthManager", () => ({
  isLoggedIn: jest.fn(),
  removeUser: jest.fn(),
}));

describe("App unit tests", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
    isAuthenticated(null);
  });

  it("Renders and sets isAuthenticated to false", async () => {
    jest.useFakeTimers();
    AuthManager.isLoggedIn.mockImplementation(
      async () => await new Promise((resolve, reject) => reject({ wow: true }))
    );
    const { asFragment } = renderWithProviders(<App />);

    expect(asFragment()).toMatchSnapshot();
    jest.runAllTimers();
    await waitFor(() => expect(isAuthenticated()).toBe(false));
  });

  it("Renders and sets isAuthenticated to true", async () => {
    jest.useFakeTimers();
    AuthManager.isLoggedIn.mockImplementation(
      async () => await new Promise((resolve) => resolve({ wow: true }))
    );
    const { asFragment } = renderWithProviders(<App />);

    expect(asFragment()).toMatchSnapshot();
    jest.runAllTimers();
    await waitFor(() => expect(isAuthenticated()).toBe(true));
  });
});
