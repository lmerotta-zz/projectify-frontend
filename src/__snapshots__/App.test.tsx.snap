// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App unit tests Renders and sets isAuthenticated to false 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
    id="react-toastify"
  />
  <div
    data-testid="root"
  />
</DocumentFragment>
`;

exports[`App unit tests Renders and sets isAuthenticated to true 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
    id="react-toastify"
  />
  <div
    data-testid="root"
  />
</DocumentFragment>
`;
